test-tests/shared/src/test/scala/zio/test/SmartAssertionSpec.scala

```diff
@@ -19,389 +20,451 @@ object SmartAssertionSpec extends ZIOBaseSpec {
+          trait Parent
+          case class Child(x: String) extends Parent
+          val someParent: Option[Parent] = Some(Child("hii"))
+          val someChild                  = Child("hii")
+          assertTrue(someParent.contains(someChild))
+        },
+        test("failure") {
+          trait Parent
+          case class Child(x: String) extends Parent
+          val someParent: Option[Parent] = None
+          val someChild                  = Child("hii")
+          assertTrue(someParent.contains(someChild))
+        } @@ failing
+      ),
+      suite("custom assertions")(
+        test("reports source location of actual usage") {
+          customAssertion("hello")
         } @@ failing
       )
-    ),
-    suite("subtype option")(
-      test("success") {
-        trait Parent
-        case class Child(x: String) extends Parent
-        val someParent: Option[Parent] = Some(Child("hii"))
-        val someChild                  = Child("hii")
-        assertTrue(someParent.contains(someChild))
-      },
-      test("failure") {
-        trait Parent
-        case class Child(x: String) extends Parent
-        val someParent: Option[Parent] = None
-        val someChild                  = Child("hii")
-        assertTrue(someParent.contains(someChild))
-      } @@ failing
-    ),
-    suite("custom assertions")(
-      test("reports source location of actual usage") {
-        customAssertion("hello")
-      } @@ failing
     )
-  )
 
   // The implicit SourceLocation will be used by assertTrue to report the
   // actual location.
   def customAssertion(string: String)(implicit sourceLocation: SourceLocation): Assert =
     assertTrue(string == "coool")
 
   // Test Types
-  private sealed trait Color
-  private final case class Red(foo: Int)     extends Color
-  private final case class Blue(bar: String) extends Color
+  sealed trait Color
+
+  object Color {
+    def asRed(color: Color): SmartAssertion[Red] = SmartAssertion.fromEither {
+      color match {
+        case Red(int) => Right(Red(int))
+        case _        => Left(s"$color IS NOT NICE!")
+      }
+    }
+  }
+
+  final case class Red(foo: Int)     extends Color
+  final case class Blue(bar: String) extends Color
+
+  implicit final class ColorOps(private val self: Color) extends AnyVal {
+    def asRed: SmartAssertion[Red] =
+      SmartAssertion.fromEither {
+        self match {
+          case Red(int) => Right(Red(int))
+          case _        => Left(s"$self IS NOT NICE!")
+        }
+      }
+  }
 
   private final case class MyClass(name: String)
   private final case class OtherClass(name: String)
+
+  implicit final class DoubleOps(private val self: Double) extends AnyVal {
+    def ~=(that: Double): SmartAssertion[Boolean] =
+      SmartAssertion.cond(self == that, s"THIS AIN'T RIGHT $self !~= $that")
+
+    def asEven: SmartAssertion[Double] =
+      SmartAssertion.fromEither(Either.cond(self % 2 == 0, self, s"$self must be even."))
+  }
```

Comment:
Another example.

Reply:
Looks great!  If we use this with case classes, should we provide `DiffResult.render` output as parameter to  `SmartAssertion.cond`?  I'm wondering how it will look in final output.